hw30_daily_task/js/main.js

```diff
@@ -0,0 +1,87 @@
+let ul = document.querySelector("#ul");
+ul.addEventListener("click", function (e) {
+  if (e.target == this) {
+    return false;
+  }
+  if (e.ctrlKey) {
+    e.target.classList.toggle("selected");
+  } else if (e.shiftKey) {
+    let selected = this.querySelector(".selected");
+    if (selected) {
+      let start = Array.from(this.children).indexOf(selected);
+      let end = Array.from(this.children).indexOf(e.target);
+      let min = Math.min(start, end);
+      let max = Math.max(start, end);
+      for (let i = min; i <= max; i++) {
+        this.children[i].classList.add("selected");
+      }
+    } else {
+      let start = Array.from(this.children).indexOf(this.firstElementChild);
+      let end = Array.from(this.children).indexOf(e.target);
+      let min = Math.min(start, end);
+      let max = Math.max(start, end);
+      for (let i = min; i <= max; i++) {
+        this.children[i].classList.add("selected");
+      }
+    }
+  } else {
```

Comment:
ok lets refactor your code here:
1. all this logic with shiftKey need to be wrapped in separate function (with name selectRange and 2 arguments in it (selected, target).
(keep in mind that `this` will be different in this case, so you'll need to replace all this with ul variable)
```javascript
} else if (e.shiftKey) {
  let selected = this.querySelector(".selected");
  selectRange(selected, target);
} else {
// ...
function selectRange(lastSelected, target) {
  if (lastSelected) {
    let start = Array.from(ul.children).indexOf(selected);
    let end = Array.from(ul.children).indexOf(target);
    let min = Math.min(start, end);
    let max = Math.max(start, end);
    for (let i = min; i <= max; i++) {
      Array.from(ul.children)[i].classList.add("selected");
    }
  } else {
    let start = Array.from(ul.children).indexOf(ul.firstElementChild);
    let end = Array.from(ul.children).indexOf(e.target);
    let min = Math.min(start, end);
    let max = Math.max(start, end);
    for (let i = min; i <= max; i++) {
      Array.from(ul.children)[i].classList.add("selected");
    }
  }
}
```
after this step you need to retest all logic to be confident it's working fine

Reply:
what else i don't like here is code doubling (Don't Repeat Yourself is one of the basic principle of programming)
2. here how you can refactor your code to prevent doubling
```javascript
function selectRange(lastSelected) {
  let startOfRange = lastSelected ? lastSelected : ul.firstElementChild;
  let start = Array.from(ul.children).indexOf(startOfRange);
  let end = Array.from(ul.children).indexOf(target);
  let max = Math.max(start, end);
  let min = Math.min(start, end);

  for (let i = min; i <= max; i++) {
    Array.from(ul.children)[i].classList.add("selected");
  }
}
```
retest after implementing this step

Reply:
but also I don't like how this loop build up.
3. lets wrap `Array.from(ul.children)[i]` in a variable, and also we have separate function addSelected so:
```javascript
for (let i = min; i <= max; i++) {
    let li = Array.from(ul.children)[i];
    addSelected(li)
}
```